CAVEATS.md

```diff
@@ -0,0 +1,175 @@
+verification is currently still in progress. For RISC-V, design-level proofs
+have completed, and C-level proofs are in progress. Similar proofs for AArch64
+are planned.
+
+The MCS configuration is supported by the seL4 foundation and should generally
+be stable, with small API changes to be expected while verification is ongoing
+and the configuration is used in more systems. See open [requests for
+comments][RFC] (RFCs) for MCS for what is currently being discussed.
+
+## SMP
+
+A symmetric multi-processor (SMP) configuration for seL4 exists and is supported
+by the seL4 foundation, but currently without formal verification. While
+generally stable, there are a small number of known open issues, in particular
+when the kernel is compiled with `clang`. We recommend `gcc` for working with
+SMP configurations of seL4.
+
+The combination of SMP and hypervisor extensions is supported and should be
+generally stable, but like the plain SMP configuration it is not formally
+verified.
+
+The combination of SMP and MCS is supported and is receiving active development,
+but it is less explored and less tested. It should still be considered
+experimental. There are no supported Armv7-a boards for SMP+MCS, only Armv8-a,
+RISC-V, and Intel. It is tested with `gcc` on `hifive`, `tqma8xqp1gb`,
+`odroidc4`, `zynqmp`, `tx1`, `tx2`, `pc99-32`, and `pc99-64`.
+
+The combination of SMP, MCS, and hypervisor extensions is currently supported on
+AArch64 only. It is less tested with lower code coverage; currently with `gcc`
+only, on `odroidc4`, `tx1`, and `tx2`.
+
+The combination of SMP and domain scheduler is not supported. The SMP
+configuration is not expected to satisfy strong intransitive non-interference
+for information flow.
+
+See the [seL4 issue tracker][issues] and the [sel4test issue tracker][sel4test
+issues] for details using the labels `MCS` and `SMP` for finding issues on these
+configurations.
+
+As these are unverified configurations, standard C implementation defects are
+possible and not excluded as in verified seL4 configurations.
+
+Supporting a static multi-kernel configuration with formal verification is on
+the roadmap for the AArch64 architecture, with initial work begun. We expect
+multi-kernel configurations to be more robust than SMP configurations, because
+they are simpler and closer to the current sequential seL4 proofs.
+
+In a multi-kernel configuration, each CPU core runs a separate instance of seL4,
+with each kernel instance getting access to disjoint subsets of memory of the
+machine. User-level memory can be shared as device-untyped memory, which the
+kernel manages but does not access. These configurations can already be set up
+without kernel changes by providing suitable device tree overlays to each kernel
+instance. Further work is planned to make such configurations easier to use and
+more robust against unsafe use/configurations, e.g. by managing IRQ controller
+access for each instance.
```

Comment:
Considering this is the caveats file, perhaps make it clearer whether current SMP will be supported in the future, or if it the plan is to obsolete it with multi-kernel. Now it hints at the latter.

Reply:
I don't actually know -- if we can fix the current issues there is no reason to remove it, but we haven't been very successful at that in the past year.

Reply:
Mostly because apparently only Kent and me do anything SMP related and we were both distracted by other things last year (Kent more than me, but I can't do anything SMP related without him).

The only advantage of current SMP are simpler system initialisation (partly because of the ability to do cross-core calls) and more efficient cross-core notifications directly to the correct task. In all other regards it's worse: Lower performance and less verifiable.

If the plan isn't to obsolete current SMP, then perhaps multi-kernel support shouldn't be mentioned in the caveats file.

Reply:
Have now clarified that multikernel is an intermediate step. Hopefully that makes it sound less like SMP is on the way to the scrap heap.